### PR TITLE
Configure agent tool round limit

### LINKED ISSUE #911: [simple-agent-poc] max_tool_rounds をエージェント定義 (YAML) に移す
## Overview / 概要

ReAct ループの上限 (`MAX_TOOL_ROUNDS = 5`) をコード定数からエージェント定義 YAML の設定フィールドに移し、エージェントごとに調整可能にする。

## Objective / 目的

`model`, `temperature`, `tools`, `stream`, `api_type` と同様に、エージェントの振る舞いを決める設定を YAML に集約する。モデルごとに適切なループ回数を調整できるようにする。

## Background / 背景

現在 `use_cases.py` に `MAX_TOOL_ROUNDS = 5` がハードコードされており、全エージェント共通。軽量モデルは 3 回で十分、推論特化モデルは 10 回欲しいといったユースケースに対応できない。`model` や `temperature` など他のエージェント挙動設定は YAML に置かれているため、一貫性の観点からも YAML 側に移すべき。

また `docs/llm-integration.md` にはエージェントループ（ReAct ループ）の概念自体が記載されておらず、LLM クライアント層とオーケストレーション層の関係性をドキュメント化する必要がある。

## Related Issues/PRs / 関連Issue・PR

None

## Implementation Approach / 実装方針

- `AgentDefinition` dataclass に `max_tool_rounds: int = 5` を追加
- YAML 未指定時はデフォルト 5 が適用され後方互換性を保つ
- バリデーションで 1〜20 の範囲に制限
- `use_cases.py` の `MAX_TOOL_ROUNDS` 定数を削除し、`agent_definition.max_tool_rounds` を参照
- 関連ドキュメント（`agent-definition.md`, `sse.md`, `session.md`, `llm-integration.md`）を更新

### YAML スキーマ追加

```yaml
agents:
  default:
    model: gpt-4.1-nano
    stream: false
    system_prompt: "..."
    temperature: null
    tools:
      - get_current_time
      - concat
      - ask_user
    api_type: completion
    max_tool_rounds: 5   # ← 新規追加（省略可能、デフォルト 5）
```

### バリデーションルール

| 条件 | 結果 |
|:---|:---|
| 未指定 / `null` | デフォルト `5` を適用 |
| `int` かつ `1 <= value <= 20` | 有効 |
| `0` 以下 | `ValidationError` |
| `21` 以上 | `ValidationError` |
| 非 `int`（`bool`, `str`, `float` 等） | `ValidationError` |

### データフロー

```mermaid
flowchart TD
    A["agents.yaml"] --> B["AgentDefinitionRegistry.from_yaml_file()"]
    B --> C["_build_agent_definition()"]
    C --> D["_optional_int_min_max(value, 1, 20, default=5)"]
    D --> E["AgentDefinition(max_tool_rounds=int)"]
    E --> F["RunAgentUseCase.execute() / execute_stream() / continue_stream()"]
    F --> G["range(agent_definition.max_tool_rounds)"]
```

#### `core/agent_definition.py`

```python
# AgentDefinition dataclass に追加
max_tool_rounds: int = 5

# _AGENT_FIELDS に追加
{"model", "system_prompt", "temperature", "tools", "api_type", "stream", "max_tool_rounds"}

# 新規バリデータ
def _optional_int_min_max(value: object, path: str, *, minimum: int, maximum: int, default: int) -> int:
    if value is None:
        return default
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValidationError(f"{path} must be an integer")
    if value < minimum or value > maximum:
        raise ValidationError(f"{path} must be between {minimum} and {maximum}")
    return value
```

#### `application/use_cases.py`

```python
# 削除
MAX_TOOL_ROUNDS = 5

# execute() L106
for _ in range(agent_definition.max_tool_rounds):  # 変更

# execute_stream() L182
for round_idx in range(agent_definition.max_tool_rounds):  # 変更

# continue_stream() L337
for round_idx in range(resume_round + 1, agent_definition.max_tool_rounds):  # 変更
```

### ドキュメント更新内容

| ファイル | 変更内容 |
|:---|:---|
| `docs/agent-definition.md` | `max_tool_rounds` フィールドの説明を追加（型、デフォルト、範囲、挙動） |
| `docs/sse.md` L138 | `MAX_TOOL_ROUNDS = 5` → `agent definition の max_tool_rounds（デフォルト 5）` |
| `docs/session.md` L26 | 同上 |
| `docs/llm-integration.md` | エージェントループ（ReAct ループ）の概念と `max_tool_rounds` による制御について追記 |

## Tasks / タスク

- [ ] `AgentDefinition` dataclass に `max_tool_rounds: int = 5` を追加
- [ ] `_AGENT_FIELDS` に `max_tool_rounds` を追加
- [ ] `_optional_int_min_max()` バリデータを実装
- [ ] `_build_agent_definition()` で `max_tool_rounds` をパース
- [ ] `use_cases.py` の `MAX_TOOL_ROUNDS` 定数を削除し、`agent_definition.max_tool_rounds` 参照に置換（3箇所）
- [ ] `agents.yaml` に `max_tool_rounds` のサンプル設定を追加
- [ ] `docs/agent-definition.md` に `max_tool_rounds` の説明を追加
- [ ] `docs/sse.md` の `MAX_TOOL_ROUNDS` 参照を更新
- [ ] `docs/session.md` の `MAX_TOOL_ROUNDS` 参照を更新
- [ ] `docs/llm-integration.md` にエージェントループの概念と `max_tool_rounds` 制御を追記
- [ ] `tests/test_agent_definition.py` に新規テストを追加（有効値、デフォルト、範囲外拒否、型エラー）
- [ ] `tests/test_use_cases.py` に明示的 `max_tool_rounds: 3` での限界値テストを追加
- [ ] `ruff check` / `ty` / `pytest` をパスさせる

## Out of Scope / スコープ外

- システム全体での絶対上限（ハードリミット）の導入。YAML の 1〜20 バリデーションのみで運用する。
- `max_tool_rounds` の動的変更（実行中にセッションごとに変える等）。

### Validation Policy / 検証方針

各ステップで `ruff check` / `ty` / `pytest` が通ることを確認する。

### Documentation / ドキュメント

- `docs/agent-definition.md`
- `docs/sse.md`
- `docs/session.md`
- `docs/llm-integration.md`

## Acceptance Criteria / 完了条件

- [ ] YAML の `max_tool_rounds` でエージェントごとにループ上限を設定できる
- [ ] 未指定時はデフォルト 5 で既存動作を維持
- [ ] 範囲外の値は起動時に `ValidationError` で拒否される
- [ ] `ruff check` が成功する
- [ ] `pytest` が成功する
- [ ] `ty` が成功する
- [ ] ドキュメントが最新化されている

## Test Perspectives / テスト観点

| Area / 観点 | What to Verify / 確認内容 | Method / 確認方法 |
| --- | --- | --- |
| Normal path / 正常系 | 有効な値（1, 5, 20）で AgentDefinition が正しく構築される | Unit test (`test_agent_definition.py`) |
| Normal path / 正常系 | 未指定 / null でデフォルト 5 が適用される | Unit test (`test_agent_definition.py`) |
| Normal path / 正常系 | 指定した max_tool_rounds でループが指定回数で打ち切られる | Unit test (`test_use_cases.py`) |
| Edge cases / 境界条件 | 0, 21 が拒否される | Unit test (`test_agent_definition.py`) |
| Edge cases / 境界条件 | bool, str, float が拒否される | Unit test (`test_agent_definition.py`) |
| Regression / 回帰 | デフォルト値 5 で既存の全テストが通過する | `pytest` |

## Notes / 補足

- `test_llm_client.py` では `AgentDefinition(...)` を直接構築しているが、`max_tool_rounds` はデフォルト値を持つため変更不要
- `conftest.py` のデフォルトフィクスチャも変更不要（省略時はデフォルト 5）
- `continue_stream()` の `resume_round + 1` からの範囲指定も同様に置換

### PR DESCRIPTION
**Pull request by u7chan**
## Issues

- Close #911

## Why

The ReAct tool-loop limit was hardcoded in `use_cases.py`, so all agents shared the same behavior regardless of model or YAML configuration.

## Summary

Move the tool-loop limit into agent definitions as `max_tool_rounds`, with a default of `5` for backward compatibility.

## Changes

- Add `AgentDefinition.max_tool_rounds` with startup validation for integer values from `1` to `20`
- Use `agent_definition.max_tool_rounds` in sync, streaming, and continue-stream ReAct loops
- Add sample YAML values, schema documentation, and LLM/session/SSE docs updates
- Add tests for defaults, valid values, invalid values, and configured loop limits

## Verification

- `uv run ruff format .` - passed
- `uv run ruff check .` - passed
- `uv run pytest tests/test_agent_definition.py tests/test_use_cases.py` - passed
- `uv run ty check .` - passed
- `uv run pytest` - passed

### CODE REVIEW

File: projects/simple-agent-poc/tests/test_use_cases.py [hunk 1]
```diff
@@ -178,6 +179,47 @@ def test_execute_exceeds_max_tool_rounds(default_agent_def, tool_registry):
...
+        {
+            "agents": {
+                "default": {
+                    "model": "test-model",
+                    "system_prompt": "You are a helpful assistant.",
+                    "max_tool_rounds": 3,
+                }
+            }
+        }
+    )
```
Comment (u7chan) [inline projects/simple-agent-poc/tests/test_use_cases.py]: [should] 💡

`execute_stream` と `continue_stream` でも設定ベースの `max_tool_rounds` を検証するテストを追加してください。

`test_execute_uses_agent_max_tool_rounds` は `execute()` のみカバーしていますが、`execute_stream` (180行目) と `continue_stream` (335行目) も同様に `agent_definition.max_tool_rounds` を参照しているため、これらのパスでも上限が正しく適用されることを確認するテストが必要です。

opencode (deepseek-v4-pro) がレビューしました 🤖

**Review by u7chan — COMMENTED** [review]
[nit] ✏️

`test_execute_exceeds_max_tool_rounds` の docstring に `MAX_TOOL_ROUNDS` とありますが、この定数は削除され `max_tool_rounds` (agent definition の設定値) に置き換えられました。docstring も新しい実装に合わせて更新することを推奨します。

例: "After max_tool_rounds (default 5), an LLMError is raised."

opencode (deepseek-v4-pro) がレビューしました 🤖

File: projects/simple-agent-poc/tests/test_use_cases.py (hunk 1 above)
Comment (u7chan) [reply projects/simple-agent-poc/tests/test_use_cases.py]: 対応しました。`execute_stream` と `continue_stream` でも `agent_definition.max_tool_rounds` が適用されることを検証するテストを追加しました。

Commit: a605347
Verification: `uv run pytest tests/test_agent_definition.py tests/test_use_cases.py tests/test_use_cases_stream.py`, `uv run ruff check tests/test_use_cases.py`, `uv run ty check .`

Comment (u7chan) [reply projects/simple-agent-poc/tests/test_use_cases.py]: 改善を確認しました。`test_execute_stream_uses_agent_max_tool_rounds` と `test_continue_stream_uses_agent_max_tool_rounds` が追加され、docstring も修正されています。問題ありません。\n\nopencode (deepseek-v4-pro) がレビューしました 🤖
